### ISSUE TITLE
Performance and memory analysis

### ISSUE THREAD

**Issue by ndt93** [HubTurbo/HubTurbo#673]
I'm doing performance and memory profiling on a fork of this repo, relevant to #424, #481 and #555. This issue is to provide any update to the dev team.

**Comment by ndt93** [comment]
I have completed the first iteration of memory profiling. The profile stats and traces file is quite large (~500MB) so I will find a way to upload it somewhere later. The profiled repo in my fork is synced with the upstream repo master branch as of the time of profiling.

The profiling is done using YourKit as the application is running on Windows 8.1 64bits on a x64 quad-core 2.13 GHz processor and 3.73 GB RAM. The Java runtime is Oracle jre 8u45. The Java VM instance is set to max 512MB and initial 40MB heap size and 256MB non-heap size. (Future profile will have same settings)

The usage scenarios involve having the existing HubTurbo repository. The rails and teammates repos are added afterwards. Different 'filtering', 'creating/removing panels', 'moving between repo', 'resizing windows' and 'clicking on issues' actions are interleaved. Garbage collection are forced between actions and generation numbers are used to mark objects allocations at different period of time.

I couldn't monitor the startup part since selenium will throws exceptions and fail if the profiler is attached to the application at the beginning.

Some observations:
- Heap memory usage is increased steadily to after each action in the usage scenario above.
- Forcing garbage collection is ineffective => either all heap objects are through strong references or there are memory leaks.
- When heap usage increases to max heap size, some sort of thrashing occurs in which garbage collection cycles happen more quickly and each takes longer time but doesn't release much memory => decrease in application overall performance.
A closer look at the traces:
- More than 80% of the heap objects are char[] which are part of String objects referred to by a field inside TurboIssue. Those strings store issues' comments.
- All TurboIssue objects are kept by strong references in memory since they are elements of the final List of TurboIssue in each Model, regardless of whether the Model is currently active and so are not removed by garbage collection.

**Comment by damithc** [comment]
Great start Thien. Keep exploring more, especially on identifying areas to improve.
Some observations from my own usage:
* I use it to monitor teammates, hubturbo, and powerpointlabs at the same time.
* I usually have one panel and the filter output is usually fewer than 10 issues.
* Immediately after restart it can take up about 500Mb.
* At times memory usage can go down to 100-200Mb, especially after a long idle period.
* However, at times it can shoot up to 900Mb at which point I usually restart HT.

**Comment by damithc** [comment]
Just now my memory usage shot up to 1.3Gb (for the same set up), after about 1 hour of usage. :-p

**Comment by ndt93** [comment]
Ok. I will try with the same setup and usage. I'm focusing on memory first since we have bigger problem with that and also my laptop cannot take the overhead of long period of cpu and threads profiling.

From my observation of live data, it's not taking up much cpu usage on average, just that cpu cycles are taken up by switching back and forth to garbage collection when memory is filling up. Also, the number of threads stays steady at about 30 after a while.

**Comment by damithc** [comment]
Another updated from my yesterday's use (same 3 projects): 
Memory usage is about 0.5Gb after start up (only one panel, less than 10 issues shown).
Opened another panel (all issues shown). shot up to 1Gb. Opened another panel. No increase in memory.
Closed the new two panels. Memory use still at 1Gb.
Left it running overnight. Gone back to 0.5Gb by the morning.
Opened two more panels. Memory use remains at 0.5Gb. 
In other words, no clear pattern :-p
But I've been changing the default project at various times, so that might be a hidden factor.

**Comment by ndt93** [comment]
Yup, the behavior is quite strange. Yesterday, when I'm watching a steady heap memory of 520mb for 30 minutes, then do go away for 15 minutes and came back to see it at 650mb, but I forgot to change the obj allocations marker so I didn't catch which objs were allocated in that time period >.<. I have some hypothesis for this behavior though.

I did a number of other profiling and some code tracking also. Gonna write a summary here later.

**Comment by ndt93** [comment]
I have uploaded some of the useful statistics that I could extracted from the profiles at this link:

https://drive.google.com/folderview?id=0B_Mq3DqIcSr3fmUyNFFtcWJqRWwyZkxKUjFYMDh6OE1vVGRJZGtGYTVYZElUNmJCYVFjMVk&usp=sharing

(The GargageCollection profile recorded collected objects during each GC invocation which is useful for obverving temporary objects allocation) 
The later profiles are recorded with VM options set to 1GB max heap and using parallel GC.

I have several observations from the profiles:
* Filtering and creating/removing panels don't seem to have any correlation with memory increase/decrease.
* There seems to be a relationship between updating repos and memory
  growth
    * The statistics from temporary objects profiles show a large number of
      temporary objects allocated during update from both backend and ui.
    * GC only collected a small percentage of temporary objects each time
      unless it needed a lot of memory for the next refresh.
* Similar to the first memory profile, the more persistent part of the heap memory is predominantly occupied by
  some large JavaFX objects and TurboIssues.

So I look more closely at the stats during update and track down the code for
updating (mainly for updating issues):
* All models are updated during a update/refresh
* Temporaries allocated at backend:
    * A new list of changed TurboIssue (In UpdateModelTask -> UpdateIssuesTask ->
      repo.GetUpdatedIssues(...))
    * New TurboIssue created for each changed Issue to be added to the list of
      exisisting TurboIssue (In UpdateModelTask -> UpdateIssuesTask ->
TurboIssue.reconcile(...))
    * new Model (UpdateModelTask)
    * new list of Serialization TurboIssues for all issues if anything in repo
      changed including metadata.
    * I didn't trace the ui code in details but there are also large number of
      objects in the ui.issuepanel pkg created during updates. They domniated
the memory for temp objects when there are no changes to the repo and less
temp issues created.
* During the update, the TaskRunner (aka ExecutorService aka ThreadPool)
  created and executed temporary *Task (aka Runnable). The
TaskRunner is also passed down to those tmp *Task objects which created its own tmp
sub *Task objects and use the TaskRunner to execute. The TaskRunner is also
passed down to those tmp sub *Task objects. Although I don't think there is an
obvious memory leak in this reading through the code, this process might create an
opportunity for such a problem.
Dev team can correct me if I misinterpret some parts or inform me of other
parts of the code that could be relevant.

I also did a quick CPU profiling. As expected, CPU time is taken mostly by
varous JavaFX rendering routines. Socket connection and stream reading
domninate in terms of invocation counts.
There are also various other areas to be improved regarding performance although I
think most of them might not result in significant improvements considering the increased complexity:

**Comment by acruis** [comment]
Good job so far.

One thing to clarify, by "GC only collected a small percentage of temporary objects each time unless it needed a lot of memory for the next refresh", when the GC does need a lot of memory for the next refresh, what is the actual percentage of temporary objects cleaned up? When it doesn't need the memory, what is the "small percentage" collected?

Also, regarding the large JavaFX objects and TurboIssues that are kept in heap memory, do they grow in size or do they maintain a constant size in memory? (i.e. are there more JavaFX objects andTurboIssues slowly being added to the heap memory as the program runs?)

In the previous traces (4 days ago), you mentioned that more than 80% of the heap objects are Strings referred to by a field inside TurboIssue, which you have identified as issues' comments. Do you mean the details of an issue (title, milestone, etc.) or do you mean the `Comment` objects contained in the `comments` fields of `IssueMetadata` objects, which is referred to by the `metadata` field in `TurboIssue`?

**Comment by ndt93** [comment]
The small percentage is 0% to 1%. Large percentage is about 10%. An example at 1 profile: memory was oscillating b/w 534MB - 537MB after each GC. At certain update, GC can collect ~30MB but then allocated additional ~50MB right after that. An indication of temporary objects not collected is the large number of long HTTP response strings, which quickly became unreachable by the GC root after an update/download but then are never collected until I close the application (about 1 hour of usage)

The large JavaFX objects (refering to the size of each individual object) don't grow in size. There are no particular large TurboIssue but if you are referring to the retained size of all TurboIssues together then they do grow in size over time. Collections of IssuePanelCard and IssuePanelSize also grow in size. However most of them are temporary or weak/soft references.

I was ambiguous there. It should refer to all of them together as the retained size of the TurboIssue (size of all objects reachable through TurboIssues). Although, most of the time the meta field doesn't contain anything. But it's a good point there about comments. I should perform more filterings that get more Comments to be created.

**Comment by damithc** [comment]
An observation from my side from leaving the app running overnight: some of the memory does get released after a long idle time. In the morning it can be as low as 200-300Mb, but goes back to the usual 500+ soon after I start using it again. For example, it was 200+ today morning and now it is 700, although the only thing i've done is marking a bunch of issues as 'read'.

**Comment by ndt93** [comment]
Yup, my guess is that the temporary, weak/soft referenced objects (created by ourselves or external libraries esp. egit.github) accumulate over each updates. I don't think there are memory leaks so those objects will get released eventually, in particular when there are no network availability for continuous updating.

The concern with the TaskRunner is that it uses an ExecutorService instantiated by `newCachedThreadPool` which we can find this description in the javadocs:

```java
public static ExecutorService newCachedThreadPool()
```
>Creates a thread pool that creates new threads as needed, but will reuse previously constructed >threads when they are available. These pools will typically improve the performance of programs >that execute many short-lived asynchronous tasks. Calls to execute will reuse previously >constructed threads if available. If no existing thread is available, a new thread will be created and >added to the pool. Threads that have not been used for sixty seconds are terminated and removed >from the cache. Thus, a pool that remains idle for long enough will not consume any resources. >Note that pools with similar properties but different details (for example, timeout parameters) may >be created using ThreadPoolExecutor constructors.
>**Returns:**
>>the newly created thread pool

The ExecutorService docs also says:
>An unused ExecutorService should be shut down to allow reclamation of its resources.
Although I think we might not want to shut it down unless we want to instantiate the threadpool again.

So one hypothesis is that if the update period is < 60s then the same threads are used again and again for updating and certain temporary references (depending on how they are managed) are retained. But when there is no updating, the threadpool releases those threads and in turn our temporary references become eligible for garbage collection.

**Comment by ndt93** [comment]
Also the description says `many short-lived asynchronous tasks` which I think targets short-lived in the sub-seconds scale e.g. thousands of parallel worker threads executing a recursive algorithm on the same threadpool.

**Comment by acruis** [comment]
The refresh period (when the program is idle) is 60 seconds, and is specified by the constant `REFRESH_PERIOD` in `UI.java`. However, a refresh is also triggered when the HubTurbo window gains focus and detects that the browser view (i.e. the HTML content of the accompanying Chrome window) has changed. Furthermore, the user can manually trigger a refresh at will (by pressing F5 or clicking on "Refresh" in the menu bar), so I don't think increasing the refresh period would help much.

**Comment by ndt93** [comment]
Ok thanks, you will take note of that.

**Comment by acruis** [comment]
I'm sorry, what?

**Comment by ndt93** [comment]
Note of the events at which refresh occur, and that increasing the REFRESH_PERIOD might not help. We are not suggesting any solution yet at the moment though.

**Comment by acruis** [comment]
Actually, I mentioned it so that you could do more profiling with the REFRESH_PERIOD increased and see if there are any changes in memory usage. Leave the application idle for some time with the adjusted refresh period, like what @damithc is doing.

With the data given, it's possible that the Tasks generated during a refresh might be the ones that keep TurboIssues from being collected here. Profiling with REFRESH_PERIOD increased helps to make sure that we can replicate the reduction of memory release when @damithc leaves the application running overnight.

So, yes, do take note of what I mentioned in the previous comment.

**Comment by ndt93** [comment]
So I did some 2 experiments regarding 1 of the hypothesis mentioned above:

1. Increase the `REFRESH_PERIOD` constant defined in ***ui.Ui.java*** to `90`

2. Construct a `ThreadPoolExecutor` directly and pass in a thread idling timeout of *10s*, *20s* or *30*. All other parameters are kept the same with a `ThreadPoolExecutor` returned by calling `newCachedThreadPool`.<br/>
E.g. To change timeout to *10s*. In ***RepoSource.java***. Replace
```java
private final ExecutorService pool = Executors.newCachedThreadPool();
``` 
with
```java
private final ExecutorService pool = new ThreadPoolExecutor(
        0, Integer.MAX_VALUE,
        10L, TimeUnit.SECONDS,
        new SynchronousQueue<Runnable>());
```

From my statistics, both methods help GC to perform more effectively. Bursts of memory allocated during startup or a refresh are generally collected after short period (2-3 min) of idling or low activity. This in turn help to keep memory usage more tractable. `TurboIssue` is no longer in the memory hotspots, so if there are still memory problems, it's likely to be somewhere else.

In one of the previous comment, I mentioned that the number of threads always stay constant but didn't think much about it. There are now activity of threads being destroyed if 1 of the above changes are made. So the thread idling period is actually < 60s since it's counted from time a thread finishes updating to the next update invocation.

In generally, method 1 and method 2 (set to 10s) tend to give the best performance. Method 2 (10s) seems to be more aggressive, so memory is reclaimed faster but also cause higher memory allocation during short period of activity where user trigger multiple refreshes.

Anyway, it could all be confirmation bias :P, so would be good if anyone else can replicate the above observations. I can create a pull request if needed.

**Comment by acruis** [comment]
The application still consumes around 800 MB after opening Rails, HubTurbo and Teammates on my system even after making the above changes :(

**Comment by ndt93** [comment]
All right! Yup, so I added rails in in addition to PowerPointLabs, Teammates and HubTurbo and did a profile. After all the repos are loaded, the memory usage is about 800MB. After that, it can drop to 600-650MB during idling period or occasionally shot back up to ~800MB.

As far as memory growth overtime are concerned, there doesn't seem to be steady increase from the initial allocation (800MB) from my observations. For example previously, with 3 repos excluding rails, the initial allocation is about 200MB but then grows steadily to max heap size (~1GB) during usage.

So just to clarify about your memory pattern: how is memory growth over time? i.e. Does it keep increasing after all repos are loaded and updated?

Anyway, beside the memory growth, reducing the persistent memory footprint (those 800MB) is another problem. TurboIssues is not likely a problem now since it accounts for only 12% (with rails) and 4% (without rails). A troubling data that emerged after I added rails is that there are up to 300MB of raw json and http response strings which are unreachable from GC root but still are not collected for the entire application lifetime >.<

**Comment by acruis** [comment]
Over time memory does not grow quickly if I leave the application alone. However, each refresh piles up some memory (that is not cleaned up after the refresh) and if I repeatedly trigger refreshes, the program memory can grow by an additional 200-300 MB (which brings the total memory usage to above 1GB).

**Comment by damithc** [comment]
For my set up, the usual memory size is about 500mb. But just now it shot up to 1.3Gb after about 10 minutes of normal usage. I restarted and now it is hovering around 500Mb again. I've noticed this a few times. There is a scenario that cause a sharp and unacceptable spike, although it doesn't happen often.

**Comment by ndt93** [comment]
Ok right! So what I suspected is that the default distribution of generations in the jvm is actually not suitable for the behavior of the application.

The hypothesis is this: the default Eden space inside the java heap is too small while we have to continuously allocate temporary json/http strings during each updates. The Survivor space is even smaller (a fraction of Eden space) which sometimes can't hold 1 of the json string object (in the case of rails for example). The effects of this is that those strings can easily be spilled into the Tenured space, making it harder to be claimed by the GC.

The problem with the default cached thread could be a direct consequence of the above hypothesis. So I read the code inside the ThreadPool and found that thread is kept alive by waiting in a loop for the next task to be available. Meanwhile, it still holds a reference to the previous task, which give that task (and its chained references) even more opportunities to be promoted to tenured space.

So I try out some stuffs with respect to this hypothesis. The timeout for the thread is reduced to 1s, meaning it can only process tasks in the same update batch. In addition, an effective way is to also tune the GC parameters, so I use these jvm options:

	-Xmx512M -XX:AdaptiveSizeDecrementScaleFactor=2 -XX:MaxHeapFreeRatio=50

Following the guidelines here:

* [Java Platform, Standard Edition HotSpot Virtual Machine Garbage Collection Tuning Guide](http://docs.oracle.com/javase/8/docs/technotes/guides/vm/gctuning/toc.html)
* [A short Primer to Java Memory Pool Sizing and Garbage Collectors](http://www.scalingbits.com/javaprimer)

The 2nd link, examples 2 and 5 seem quite relevant. Again, it would be nice if anyone could try the above steps out (reduce timeout and change vm options) and observe how the memory behaves :). The options target parallel GC (which is the default if your comp have > 1 core except if it's windows 32 bits running on 2 cores), but should be fine for serial GC too.

**Comment by ndt93** [comment]
Bumping this thread again.

I have done some profiles on the updated master. The memory footprint of the program is quite tamed now (about 400-450MB) for the 3 repos (HubTurbo, PowerPointLabs and Teammates), which I'm not sure is because of one of the fixes or updated libraries. In general, all objects' types are quite evenly distributed without many long-living large http/json strings as mentioned before.

Anyway, the VM options mentioned above help to limit the memory usage further down to about 300-340MB from my observations. Since these experiments are not really fully arguable logic but rather speculative (since I haven't found any definite memory bug for a clear fix), it would be helpful if anyone can replicate these observations (not necessary with a proper profile and all, just task manager).

If anyone is concerned with integrating VM options in case we actually want it, there is this java wrapper [Launch4j](http://launch4j.sourceforge.net/) which can help and also comes with other goodies. It also has a [Gradle plugin](https://github.com/TheBoegl/gradle-launch4j) and Linux package which we can use to resolve dependency for Travis CI.

**Comment by acruis** [comment]
Forcing garbage collection seems to be reliably reducing old gen pool size to <50MB now. (with Teammates and HT open)

Surprisingly, I checked out the V2.8.0 release (083a03e) (June 5) and managed to achieve the same results, too (both current master and V2.8.0 run without VM arguments).

**Comment by ndt93** [comment]
Ok that sounds good. Can let me know where you force the garbage collection so I can try out later?

**Comment by acruis** [comment]
Where in interface: there's a Force Garbage Collection button in YourKit, I pressed it.
Where in program run: I spammed <kbd>F5</kbd>, then waited a while until all the refreshes complete (the most recent log message should be a "Triggered ModelUpdatedEvent"). Then I clicked the Force Garbage Collection button.

I might not have refreshed the libraries though, I'm trying with the release .jar now.

**Comment by ndt93** [comment]
Oh right. I suppose we either have to use VM options or force garbage collection in the code for a fix though. The popular opinion is that forcing GC is generally [not recommended](https://www-01.ibm.com/support/knowledgecenter/SSYKE2_7.0.0/com.ibm.java.zos.71.doc/diag/understanding/mm_gc_manual.html)

**Comment by acruis** [comment]
Yup, I wasn't gonna put something like `System.gc()` into the code.

With the V2.8.0 .jar, I can still reduce old gen pool size to <50MB using forced garbage collection, so it's probably not thanks to updated libraries.

Does this mean that there were not many strong references in the first place?

**Comment by ndt93** [comment]
Ok then I suppose some of the new changes in the program helps out somehow :P. Yup, there are not many strong references from what I saw, they are those javafx stuffs and a loaded issues. Most of the memory are occupied by http/json strings which I think are from egit download or gson processing, and for some reasons the GC don't feel the need to clean them up...

Would be nice to improve the memory some more though since it's kind of strange for a utility like this to occupy > 400MB while the raw size of the repos are only a few MB.

**Comment by acruis** [comment]
A simple JavaFX application is about 170 MB from startup. I'm not sure if the IssuePanelCells that we generate every refresh are worth the other ~300MB :(

I think GC will never autonomously clean up, unless the eden space is filled beyond its threshold.

**Comment by ndt93** [comment]
Ok I see. I should investigate the front-end side a bit more then.

**Comment by damithc** [comment]
I hit 1.2Gb just now and seems to be staying there. The reason could be that I changed switched between two boards.
One board has one panel: `(repo:hubturbo/hubturbo OR repo:teammates/repo OR repo:powerpointlabs/powerpointlabs) updated:10 is:unread -sort:nonSelfUpdate ` Total issues shown is less than 10
Other  board has two: `milestone:V2.5.5 repo:powerpointlabs/powerpointlabs` and `milestone:V2.5.6 repo:powerpointlabs/powerpointlabs` Total issues shown is less than 10

**Comment by damithc** [comment]
Restarted, and now it's at 1.6Gb. So having multiple boards is definitely making the memory situation untenable. I'll open another issue for this.

**Comment by ndt93** [comment]
Ok, that's great. I can try this case tonight. Quick update: I did some profiles and pay more attention to the UI code this time, there doesn't seem to be anything abnormal about JavaFX stuffs themselves though, only take < 10% of total memory.

**Comment by damithc** [comment]
Looks like this memory explosion I observed was caused by a json explosion, which is not related to memory performance issue. So you can ignore my observation above.

**Comment by ndt93** [comment]
Memory usage has gotten a bit worse from the profiling results of today master's merge, varying between 550MB to 600MB and could shot up to 700MB after an `updated` filter. However, the trend remains similar to previous runs, in which memory is mostly occupied by temporary http and json strings; only 25% (<50MB) are occupied by strongly reachable objects.
